### ISSUE TITLE
Compatibility with Tpetra in Trilinos 13.4.0 / 13.2.0

### ISSUE THREAD

**Issue by tamiko** [dealii/dealii#13972]
We currently deploy the following sanity check for Tpetra:
```
#include <Tpetra_Vector.hpp>

int main() {
  using LO = int;
  using GO = unsigned int;
  using Node = Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial>;
  using map_type = Tpetra::Map<LO, GO, Node>;
  Teuchos::RCP<const map_type> dummy_map = Teuchos::rcp(new map_type());
  Tpetra::Vector<double, LO, GO, Node> dummy_vector(dummy_map);
  (void)dummy_vector;
  return 0;
}
```
which leads to a linker error on my systems because none of the Tpetra types are instantiated for `GO = unsigned int`, but for `GO = long long`. Similarly, if Trilinos happens to be compiled with openmp support then everything is instantiated for `Kokkos::OpenMP` and not `Kokkos::Serial`.

In order to fix at least the second issue (selection of `Kokkos::<backend>`) - at no point in our code do we actually force a certain backend. So would it be admissible to change the test into something like this?
```
#include <Tpetra_Vector.hpp>

int main() {
  using LO = int;
  using GO = unsigned int;
  using map_type = Tpetra::Map<LO, GO>;
  Teuchos::RCP<const map_type> dummy_map = Teuchos::rcp(new map_type());
  Tpetra::Vector<double, LO, GO> dummy_vector(dummy_map);
  (void)dummy_vector;
  return 0;
}
```

Also, `GO` should be set to our global degrees of freedom type and not to `unsigned int`.

Regarding the `GO` type - I think this is something I can set in the Trilinos configuration, so that is an issue on my side and the test correctly identifies the issue. So I would leave this as is. Alternatively, would it be possible to have an index translation from 32 bit to 64 bit for Tpetra?

**Comment by Rombur** [comment]
By default Trilinos uses `long long` for `GO`, you can enable `unsigned int` or `unsigned long long` at configure time. 

If you don't set the node, the backend  will be chosen in the following order, CUDA, OpenMP, and finally Serial. The new test should work as long as you don't compile with CUDA (or another GPU) backend. If CUDA is enabled, we need to make sure that we use the CXX compiler provided by the user. Kokkos provide a script: `nvcc_wrapper` which takes care of all the CUDA flags.

**Comment by tamiko** [comment]
@Rombur The problem is that configuring with openmp does instantiate the backend for `Kokkos::OpenMP` and not `Kokkos::Serial`, meaning the test will fail.

**Comment by Rombur** [comment]
In Kokkos there is the backend `DefaultHostSpace` which is either OpenMP or Serial depending on how Kokkos was configured. Let me look if Tpetra has something similar

**Comment by Rombur** [comment]
> The problem is that configuring with openmp does instantiate the backend for Kokkos::OpenMP and not Kokkos::Serial, meaning the test will fail.

You can see [here](https://github.com/trilinos/Trilinos/blob/81f76193e6e42796da076fe4bed906c647515ae2/packages/tpetra/classic/NodeAPI/Kokkos_DefaultNode.hpp#L71-L85) how it works. If you don't set the node, Tpetra just goes down the list of backend and then, it picks the first one that is enable which could be a GPU backend.

It looks like Tpetra does not expose `DefaultHostSpace` but I guess you can do like them and use macros to decide which backend to use (see [here](https://github.com/trilinos/Trilinos/blob/81f76193e6e42796da076fe4bed906c647515ae2/packages/tpetra/classic/NodeAPI/KokkosCompat_ClassicNodeAPI_Wrapper.hpp#L64-L74)). The `KOKKOS_ENABLE_XXX` are exposed by Kokkos

**Comment by tamiko** [comment]
ll, let me summarize my current understanding of the Tpetra change (either in version 13.2 or 13.4).

Tpetra has gained (compile time) support to choose a different global index type. During configure you will see something like this:
```
-- Tpetra: Tpetra_INST_INT_LONG_LONG is enabled by default.
-- Tpetra: Tpetra_INST_INT_UNSIGNED is disabled by default.
-- Tpetra: Tpetra_INST_INT_UNSIGNED_LONG is disabled by default.
-- Tpetra: Tpetra_INST_INT_INT is disabled by default.
-- Tpetra: Tpetra_INST_INT_LONG is disabled by default.
```
This implies that simply compiling Trilinos with instantiate Tpetra with `long long` (which is equivalent to `std::int64_t`), which does neither match our 32bit nor our 64bit global index type, which are both unsigned:
```
#ifdef DEAL_II_WITH_64BIT_INDICES
  using global_dof_index = std::uint64_t;
#else
  using global_dof_index  = unsigned int;
#endif
```

This is an issue because it seems that one can currently not compile Trilinos with either `Tpetra_INST_INT_UNSIGNED` or `Tpetra_INST_INT_UNSIGNED_LONG`. Internally portions of Trilinos require signed integers:
```
kokkos-kernels/src/sparse/KokkosSparse_CrsMatrix.hpp:377:36: error: static assertion failed: CrsMatrix requires that OrdinalType is a signed integer type.
  377 |       std::is_signed<OrdinalType>::value,
      |                                    ^~~~~
```

So as it is, we don't have Tpetra support with Trilinos 13.4 (or 13.2, depending on where the change happened) onwards.

**Comment by tamiko** [comment]
Looking at some instances how we use Tpetra at the moment I think a possible fix could be to replace all instances of `Tpetra::Import<int, types::global_dof_index>` and similar uses by something like `Tpetra::Impor<int, types::tpetra_global_dof_index>` which will lead to a number of unsigned to signed integer casts.

**Comment by Rombur** [comment]
We do have support for Trilinos 13.2, I use it all the time. 

So just a little of background, Trilinos 13.4 is the last 13.x version of Trilinos. The next version is Trilinos 14.0 and, last I heard, a _lot_ of code will be removed (the Epetra stack will probably be removed). The Tpetra interface will probably be changed too and so, the default template instantiation has been changed to get ready for Tpetra 14.0 (I think some template parameters will disappear and they won't support `unsigned int` for some template parameters). We should really try to support Trilinos 13.4 because a lot of people will be stuck with it for a while. I can take a look at it tomorrow.

**Comment by tamiko** [comment]
Short update, the missing `Kokkos::Serial` instantiations is an easy fix, just make sure that `-DTpetra_INST_SERIAL=ON` is around unconditionally.

**Comment by tamiko** [comment]
Reassigning to the 9.5 milestone.

**Comment by jpthiele** [comment]
As I just encountered this as well I had a quick look at Epetra and our wrappers.
If I am not completely mistaken we just ignored that Epetra was signed in the global ordinate. 
But I'd be happy to be corrected on this. 
Luckily we only have the vector class implemented so it's not too much to change. 
However I think that for Tpetra we should actually cast as @tamiko suggested. 

As to why unsigned int is not supported I was told that some negative ints are used for encoding additional information 
like indices being off rank.
